### PR TITLE
[PoC] Low Latency Participant [DPP-947]

### PR DESCRIPTION
**Pull request by tudor-da**
Significant changes in this PR:
* Mutable contract state cache is updated in batches, where one batch is a transaction (i.e. per offset updates). Hence, the mutable contract state cache doesn't use the event sequential id internally for the state caches indexes and for `validAt` lookups on cache misses.
* The `BufferedCommandCompletionsReader` is implemented for serving the completions stream out of the in-memory fan-out.
* The indexer is extended with a Akka Stream flow (see `BatchingParallelIngestionPipe`) that updates the Ledger API caches: in-memory fan-out buffers for transactions and completions, the mutable contract state cache, the Ledger API ledger end and general Ledger API offset dispatcher. 
* The indexer's (see `BatchingParallelIngestionPipe`) `groupedWithin` batching stage is replaced with a latency-optimized stage `batchN` that ensures constant size batch forming under load and no latency in no load scenarios.
* Bounded memory usage for fetching from the `EventsBuffer` (bound configurable per buffer). The result of slices don't keep reference the underlying buffer data structure anymore.
* Various small performance optimizations in the in-memory fan-out logic.
* One fix for buffers pruning assertion (observable as a conformance test assertion change)

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 1]
```diff
@@ -235,146 +219,17 @@ private[platform] class MutableCacheBackedContractStore(
...
-    metrics.daml.execution.cache.indexSequentialId
-      .updateValue(event.eventSequentialId)
-    event match {
-      case LedgerEndMarker(eventOffset, eventSequentialId) =>
-        signalNewLedgerHead(eventOffset, eventSequentialId)
-      case _ => ()
-    }
-  }
-
-  private val updateCaches: ContractStateEvent => Unit = {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:256]: extracted into separate class `MutableContractStateCaches`

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 2]
```diff
@@ -235,146 +219,17 @@ private[platform] class MutableCacheBackedContractStore(
...
-  ): MutableCacheBackedContractStore =
-    new MutableCacheBackedContractStore(
-      metrics,
-      contractsReader,
-      signalNewLedgerHead,
-      ContractKeyStateCache(startIndexExclusive, maxKeyCacheSize, metrics),
-      ContractsStateCache(startIndexExclusive, maxContractsCacheSize, metrics),
-    )
-
-  final class CacheUpdateSubscription(
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:350]: Not needed anymore.

File: ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/IndexerStabilityTestFixture.scala [hunk 3]
```diff
@@ -94,12 +107,34 @@ object IndexerStabilityTestFixture {
...
+                          dbDispatcher.executeSql(
+                            metrics.daml.index.db.loadStringInterningEntries
+                          ) {
+                            stringInterningStorageBackend.loadStringInterningEntries(
+                              fromExclusive,
+                              toInclusive,
+                            )
+                          }
+                    ),
+                    updateInMemoryBuffersFlow = updateInMemoryBuffersFlow,
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/IndexerStabilityTestFixture.scala:137]: Adds in-memory fan-out and mutable contract state caches load to the indexer under benchmark

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala [hunk 4]
```diff
@@ -216,40 +220,46 @@ object ParallelIndexerSubscription {
...
-        lastStringInterningId = curr.lastStringInterningId,
-        lastRecordTime = curr.lastRecordTime,
-        batch = zeroDbBatch, // not used anymore
-        batchSize = 0, // not used anymore
-        offsets = Vector.empty, // not used anymore
-      )
+  ): (WithUpdatesBatch[Batch[DB_BATCH]], WithUpdatesBatch[Batch[DB_BATCH]]) => WithUpdatesBatch[
+    Batch[DB_BATCH]
+  ] = { case ((prevUpdates, _), (currUpdates, currentBatch)) =>
+    currUpdates.copy(_1 = prevUpdates._1 ++ currUpdates._1) -> Batch[DB_BATCH](
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala:253]: Hazardous if downstream slows unexpectedly (OOM). Apply backpressure

File: ledger/sandbox-on-x/BUILD.bazel [hunk 5]
```diff
@@ -479,12 +479,12 @@ conformance_test(
...
 
 conformance_test(
-    name = "conformance-test-in-memory-fan-out",
+    name = "conformance-test-in-memory-fan-out-small-caches",
     ports = [6865],
     server = ":app",
     server_args = [
         "--contract-id-seeding=testing-weak",
-        "--participant participant-id=example,port=6865,ledger-api-transactions-buffer-max-size=10",
+        "--participant participant-id=example,port=6865,contract-state-cache-max-size=10,contract-key-state-cache-max-size=3,ledger-api-transactions-buffer-max-size=3",
```
Comment (tudor-da) [inline ledger/sandbox-on-x/BUILD.bazel:487]: Test exercising boundary conditions for in-memory fan-out and mutable state caches.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala [hunk 6]
```diff
@@ -180,19 +186,31 @@ private[platform] object BufferedTransactionsReader {
...
-        // Using collect + mapAsync as an alternative to the non-existent collectAsync
-        .collect { case (offset, tx: TxUpdate) =>
-          Timed.future(toApiTxTimer, toApiTx(tx, filter, verbose).map(offset -> _))
+        .fromIterator(() =>
+          slice
+            .collect { case (offset, tx: TransactionLogUpdate.TransactionAccepted) =>
+              offset -> tx
+            }
+            .iterator
+            .grouped(10)
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala:201]: Bit magical

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala:201]: Missed this one. The purpose here was to reduce Futures/Akka tripping. Good observation, it won't make it like this int the prod PR.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala [hunk 7]
```diff
@@ -85,12 +89,13 @@ private[platform] case class StateCache[K, V](
...
     */
-  def putAsync(key: K, fetchAsync: Long => Future[V])(implicit
+  def putAsync(key: K, fetchAsync: Offset => Future[V])(implicit
       loggingContext: LoggingContext
   ): Future[V] = Timed.value(
     registerUpdateTimer,
     pendingUpdates.synchronized {
-      val validAt = cacheIndex
+      // TODO LLP: Revisit sanity of this condition
+      val validAt = cacheIndex.getOrElse(Offset.fromHexString(Ref.HexString.assertFromString("FF")))
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala:98]: Looks very wrong.

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala:98]: That's why I added the TODO :) Cut some corners for the PoC where this edge case doesn't matter that much and it does the trick

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala [hunk 8]
```diff
@@ -117,16 +117,6 @@ object ParallelIndexerFactory {
             _ <- meteringAggregator(dbDispatcher)
           } yield dbDispatcher
         ) { dbDispatcher =>
-          val stringInterningView = new StringInterningView(
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala:120]: unfortunately there was a reason why we instantiated the view for each indexing session: this view is being populated ahead (interning entries which are not persisted yet), which means we need to "roll it back", as we restart indexing. This was achieved with reinitialization. We probably need to make interning smarter to support this use case here (the same interning for both indexer and index)

File: ledger/participant-integration-api/src/main/scala/platform/index/ParticipantInMemoryState.scala [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    val transactionsBuffer: EventsBuffer[TransactionLogUpdate],
+    val completionsBuffer: EventsBuffer[TransactionLogUpdate],
+    updateLedgerApiLedgerEnd: LedgerEnd => Unit,
+    metrics: Metrics,
+) {
+  private val logger = ContextualizedLogger.get(getClass)
+  private val buffersUpdaterExecutionContext =
+    ExecutionContext.fromExecutorService(Executors.newSingleThreadExecutor())
+
+  def updateBatch(update: TransactionLogUpdate)(implicit
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/ParticipantInMemoryState.scala:31]: adding the update logic to the class with values causing the usual oo friction in factories: the pulling up of the jdbcledgerdao and friends is a consequence that the updateLedgerApiLedgerEnd needs to be instantiated before StandAloneIndexService. Separating data from logic here might allow to put them back

**Review by nmarton-da — COMMENTED** [review]
nice progress!

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala [hunk 10]
```diff
@@ -229,4 +247,18 @@ private[platform] object BufferedTransactionsReader {
       size = outputStreamBufferSize,
     )
   }
+
+  private implicit class SourceWithBuffers[T, R](source: Source[T, NotUsed]) {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala:277]: I am aware of that. For the sake of readability in this PoC, I preferred this approach, fluent API style.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala [hunk 11]
```diff
@@ -42,9 +47,34 @@ object BatchingParallelIngestionPipe {
       .mapAsync(ingestingParallelism)(ingester)
       // Stage 6: Preparing data sequentially for throttled mutations in database (tracking the ledger-end, corresponding sequential event ids and latest-at-the-time configurations)
       .conflate(tailer)
-      .throttle(tailingRateLimitPerSecond, FiniteDuration(1, "seconds"))
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:45]: Isn't the conflate doing just that?

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala [hunk 12]
```diff
@@ -96,6 +97,7 @@ private[platform] case class ParallelIndexerSubscription[DB_BATCH](
 }
 
 object ParallelIndexerSubscription {
+  type WithUpdatesBatch[T] = ((Iterable[(Offset, state.Update)], LedgerEndMarker), T)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala:100]: I could have but I was thinking it's easier to review like this.

File: ledger/ledger-on-memory/BUILD.bazel [hunk 13]
```diff
@@ -41,54 +40,54 @@ da_scala_library(
...
-        "//libs-scala/contextualized-logging",
-        "//libs-scala/resources",
-        "//libs-scala/resources-akka",
-        "//libs-scala/resources-grpc",
-        "@maven//:io_dropwizard_metrics_metrics_core",
-        "@maven//:org_mockito_mockito_core",
-        "@maven//:org_scalatest_scalatest_compatible",
-    ],
-)
+#da_scala_test_suite(
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/ledger-on-memory/BUILD.bazel:43]: Not this in this PoC, exactly, but the changes to the `SandboxOnXRunner` should be translatable here as well.

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ParticipantPruningIT.scala [hunk 14]
```diff
@@ -131,7 +131,7 @@ class ParticipantPruningIT extends LedgerTestSuite {
         LedgerApiErrors.RequestValidation.ParticipantPrunedDataAccessed,
         Some(
           Pattern.compile(
-            s"(Transactions request from [0-9a-fA-F]* to [0-9a-fA-F]* precedes pruned offset ${offsetToPruneUpTo.getAbsolute})|(Request from [0-9a-fA-F]* precedes pruned offset ${offsetToPruneUpTo.getAbsolute})"
+            s"(request from [0-9a-fA-F]* to [0-9a-fA-F]* precedes pruned offset ${offsetToPruneUpTo.getAbsolute})|(Request from [0-9a-fA-F]* precedes pruned offset ${offsetToPruneUpTo.getAbsolute})"
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ParticipantPruningIT.scala:134]: Yes. There was a bug in the prune safe-guard in the in-memory fan-out which, when fixed, shows a slightly different error message towards the client. I think it can be unified though with the DB-raised error so we can remove the need for this adaptation here.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 15]
```diff
@@ -352,6 +354,13 @@ final class Metrics(val registry: MetricRegistry) {
     object index {
       private val Prefix = daml.Prefix :+ "index"
 
+      object IndexBypassBuffer {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:357]: No 👍

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala (hunk 8 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala:120]: Why do we need to roll it back? Are we expecting different incoming data from the updates stream? As I see insertion into the StringInterningView is idempotent (https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/interning/RawStringInterning.scala#L31)

File: ledger/participant-integration-api/src/main/scala/platform/index/ParticipantInMemoryState.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/index/ParticipantInMemoryState.scala:31]: Yes, that's easy to do. Stopped here with the batching for convenience though

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala [hunk 16]
```diff
@@ -180,19 +186,31 @@ private[platform] object BufferedTransactionsReader {
...
+          slice
+            .collect { case (offset, tx: TransactionLogUpdate.TransactionAccepted) =>
+              offset -> tx
+            }
+            .iterator
+            .grouped(10)
+        )
+        .collectType[Seq[(Offset, TransactionLogUpdate.TransactionAccepted)]]
+        .buffered(32)(inStreamBufferLength)
+        .mapAsync(2) { batch =>
```
[the remark this answers is not in the document]
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/BufferedTransactionsReader.scala:205]: wow!

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 17]
```diff
@@ -87,7 +85,17 @@ private[platform] final class EventsBuffer[O: Ordering, E](
     * @param endInclusive The end inclusive bound of the requested range.
     * @return The series of events as an ordered vector satisfying the input bounds.
     */
-  def slice(startExclusive: O, endInclusive: O): BufferSlice[(O, E)] =
+  def slice(startExclusive: Offset, endInclusive: Offset): BufferSlice[(Offset, E)] = if (
+    lastPrunedOffset.exists(_ > startExclusive)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala]: Because it was missing
